packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -288,48 +264,168 @@ class CupertinoDialogAction extends StatelessWidget {
+          child: new Column(
+            mainAxisSize: MainAxisSize.max,
+            crossAxisAlignment: CrossAxisAlignment.stretch,
+            children: titleContentGroup,
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+
+/// A Action Items section typically used in a [CupertinoAlertDialog].
+///
+/// If only two or less action are present, they are laid out horizontally. If
+/// more than two actions are present, they are laid out vertically in a single
+/// column. If there isn't enough room to show all the buttons, they are
+/// wrapped in a [CupertinoScrollbar] widget. If children is null or empty,
+/// it returns null.
+class _CupertinoAlertActionSection extends StatelessWidget {
+  const _CupertinoAlertActionSection({
     Key key,
     this.children,
+    this.scrollController,
   }) : super(key: key);
 
   final List<Widget> children;
 
+  /// A scroll controller that can be used to control the scrolling of the message
+  /// in the dialog.
+  ///
+  /// Defaults to null, and is typically not needed, since most alert messages are short.
+  final ScrollController scrollController;
+
   @override
   Widget build(BuildContext context) {
+    if (children == null || children.isEmpty) {
+      return null;
+    }
+
     final List<Widget> buttons = <Widget>[];
 
     for (Widget child in children) {
       // TODO(abarth): Listen for the buttons being highlighted.
-      // TODO(gspencer): These buttons don't lay out in the same way as iOS.
-      // When they get wide, they should stack vertically instead of in a row.
-      // When they get really big, the vertically-stacked buttons should scroll
-      // (separately from the top message).
       buttons.add(new Expanded(child: child));
     }
 
-    return new CustomPaint(
-      painter: new _CupertinoButtonBarPainter(children.length),
-      child: new UnconstrainedBox(
-        constrainedAxis: Axis.horizontal,
-        child: new ConstrainedBox(
-          constraints: const BoxConstraints(minHeight: _kButtonBarHeight),
-          child: new Row(children: buttons),
+    if (_shouldLayoutActionsVertical(children.length)) {
+      return new Flexible (flex: 1,
+          child: new CupertinoScrollbar(
+            child: new SingleChildScrollView(
+              controller: scrollController,
+              child: new CustomPaint(
+                painter: new _CupertinoVerticalActionPainter(buttons.length),
+                child: new ConstrainedBox(
+                  constraints: new BoxConstraints(maxHeight: _kButtonBarHeight * buttons.length),
+                  child: new Column(
+                      mainAxisSize: MainAxisSize.max,
+                      crossAxisAlignment: CrossAxisAlignment.stretch,
+                      children: buttons
+                  ),
+                ),
+              ),
+            ),
+          ),
+        );
+    } else {
+      return new CustomPaint(
+        painter: new _CupertinoHorizontalActionPainter(children.length),
```

Comment:
I'd keep the word divider or something in the name, since it's not painting the actions but the action divider